pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+		clusterCIDRSynced:     clusterCIDRInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel: make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:              recorder,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "multi_cidr_range_allocator"),
+		lock:                  &sync.Mutex{},
+		nodesInProcessing:     sets.NewString(),
+		cidrMap:               make(map[string][]*cidrset.ClusterCIDR, 0),
+	}
+
+	// testCIDRMap is only set for testing purposes.
+	if len(testCIDRMap) > 0 {
+		ra.cidrMap = testCIDRMap
+		klog.Warningf("testCIDRMap should only be set for testing purposes, if this is seen in production logs, it might be a misconfiguration or a bug.")
+	}
+
+	ccList, err := listClusterCIDRs(client)
+	if err != nil {
+		return nil, err
+	}
+
+	if ccList == nil {
+		ccList = &networkingv1alpha1.ClusterCIDRList{}
+	}
+	createDefaultClusterCIDR(ccList, allocatorParams)
+
+	// Regenerate the cidrMaps from the existing ClusterCIDRs.
+	for _, clusterCIDR := range ccList.Items {
+		klog.Infof("Regenerating existing ClusterCIDR: %v", clusterCIDR)
+		// Create an event for invalid ClusterCIDRs, do not crash on failures.
+		if err := ra.reconcileBootstrap(&clusterCIDR); err != nil {
+			klog.Errorf("Error while regenerating existing ClusterCIDR: %v", err)
+			ra.recorder.Event(&clusterCIDR, "Warning", "InvalidClusterCIDR encountered while regenerating ClusterCIDR during bootstrap.", err.Error())
+		}
+	}
+
+	clusterCIDRInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    createClusterCIDRHandler(ra.reconcileCreate),
+		DeleteFunc: createClusterCIDRHandler(ra.reconcileDelete),
+	})
+
+	if allocatorParams.ServiceCIDR != nil {
+		ra.filterOutServiceRange(allocatorParams.ServiceCIDR)
+	} else {
+		klog.V(0).Info("No Service CIDR provided. Skipping filtering out service addresses.")
+	}
+
+	if allocatorParams.SecondaryServiceCIDR != nil {
+		ra.filterOutServiceRange(allocatorParams.SecondaryServiceCIDR)
+	} else {
+		klog.V(0).Info("No Secondary Service CIDR provided. Skipping filtering out secondary service addresses.")
+	}
+
+	if nodeList != nil {
+		for _, node := range nodeList.Items {
+			if len(node.Spec.PodCIDRs) == 0 {
+				klog.V(4).Infof("Node %v has no CIDR, ignoring", node.Name)
+				continue
+			}
+			klog.V(0).Infof("Node %v has CIDR %s, occupying it in CIDR map", node.Name, node.Spec.PodCIDRs)
+			if err := ra.occupyCIDRs(&node); err != nil {
+				// This will happen if:
+				// 1. We find garbage in the podCIDRs field. Retrying is useless.
+				// 2. CIDR out of range: This means ClusterCIDR is not yet created
+				// This error will keep crashing controller-manager until the
+				// appropriate ClusterCIDR has been created
+				return nil, err
+			}
+		}
+	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controllerutil.CreateAddNodeHandler(ra.AllocateOrOccupyCIDR),
```

Comment:
I think that this is taking the lock during the whole operation, means we are serializing all the events

Reply:
Would be the same behavior here: https://github.com/kubernetes/kubernetes/blob/137439194d8e27f34081218a7e8cfb1e31e74a69/pkg/controller/nodeipam/ipam/range_allocator.go#L137-L138

Reply:
Yeah, I see now that we are carrying over some legacy behaviour that I can have an opinion since I never used it, maybe @bowei have more context about the stability of that controller